File: src/main/kotlin/web/url/UrlForTracking.kt
package web.url

class UrlForTracking(
    private val original: Url,
    url: String,
) : AbstractUrl(url, original.getStrategy()) {
    override fun getOriginalUrl() = original.getOriginalUrl()
    override fun recordSuccessfulFetch() = original.recordSuccessfulFetch()
    override fun recordFailedFetch(code: Int, reason: String) = original.recordFailedFetch(code, reason)
}

File: symfony/src/Utils/Web/Url/UrlForTracking.php
<?php

declare(strict_types=1);

namespace App\Utils\Web\Url;

use App\Utils\Web\UrlStrategy\Strategy;
use Override;

readonly class UrlForTracking extends AbstractUrl
{
    public function __construct(
        string $url,
        public Url $original,
        ?Strategy $strategy = null,
    ) {
        parent::__construct($url, $strategy);
    }

    #[Override]
    public function getOriginalUrl(): string
    {
        return $this->original->getUrl();
    }

    #[Override]
    public function recordSuccessfulFetch(): void
    {
        $this->original->recordSuccessfulFetch();
    }

    #[Override]
    public function recordFailedFetch(int $code, string $reason): void
    {
        $this->original->recordFailedFetch($code, $reason);
    }
}
